cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,305 @@
+    result = solveset(u, V, domain=S.Reals)
+    # The resul could be an intersection with Reals, if a custom exp function is used. We assume the result is real.
+    if isinstance(result, Intersection) and result.args[0] == S.Reals:
+        result = result.args[1]
+    return result
+
+
+def _get_singularity(expr, V, U_offset, exp_function):
+    """
+    Finds singularities in equations of the form:
+       - ``U / (exp(U) - 1.0)``
+       - ``U / (1.0 - exp(U))``
+       - ``(exp(U) - 1.0) / U``
+       - ``(1.0 - exp(U)) / U``
+
+    In addition to the singularity itself, this functions returns a lower and upper bounds as follows:
+    - sp is the singularity point: the value of V for which U == 0
+    - Vmin is the value of V for which U - U_offset == 0
+    - Vmax is the value of V for which U + U_offset == 0:
+    :return: (Vmin, Vmax, sp)
+    """
+
+    # Create "wildcards", that act as catch-all (*) when matching expressions
+    # https://docs.sympy.org/latest/modules/core.html#sympy.core.basic.Basic.match
+    P = Wild('P', real=True, exclude=[V])
+    Z = Wild('Z', real=True)
+    U_wildcard = Wild('U_wildcard', real=True, include=[V])
+    SP_wildcard = Wild('SP_wildcard', real=True)
+    singularities = []
+
+    def check_U_match(m, sp):
+        assert m is None or m[Z] != 0
+        return m is not None and \
+            (sp == m[SP_wildcard] or (isinstance(sp, Float) and
+             isinstance(m[SP_wildcard], Float) and isclose(m[SP_wildcard], sp)))
+
+    # find all fractions and sperate numerator and denominator
+    # the denominator is all args where a **-1
+    numerator = [a for a in expr.args if not _is_negative_power(a)]
+    denominator = [Pow(a.args[0], - a.args[1]) for a in expr.args if _is_negative_power(a)]
+
+    # Not a division or does not have exp
+    if len(denominator) == 0 or len(numerator) == 0 or not expr.has(exp_function):
+        return []
+
+    numerator += [Mul(*numerator)]  # pattern could match entire numerator as well as any of its parts
+    denominator += [Mul(*denominator)]  # pattern could match entire denominator as well as any of its parts
+
+    # U might be on top, try numerator / denominator and denominator / numerator
+    for fraction_part_1, fraction_part_2 in ((numerator, denominator), (denominator, numerator)):
+        found_on_top = False
+        for fp2 in fraction_part_2:  # Check arguments in denominator (or numerator)
+            (Vmin, Vmax, sp) = None, None, None
+            if not fp2.has(exp_function):
+                continue
+            # look for exp(U) * -Z + 1.0
+            # Works as we later reuire Z to be positive
+            find_U = fp2.match(exp_function(U_wildcard) * -Z + 1.0)
+            if not find_U:
+                find_U = fp2.match(exp_function(U_wildcard) * Z - 1.0)  # look for exp(U) * Z - 1.0
+
+            # Z should be positive, we replace free symbols by 1 to be able to evaluate the sign of Z
```

Comment:
I don't understand this step. Why can't we evalf() Z to get its sign? Are free symbols just variables? And how do we know they are positive so that setting them to 1 is OK?

Reply:
free_symbols are just variables yes.

Z could contain a parameter. However it's a fair point this method only checks that there isn't a negative sign in this expression but the variables could represent negative numbers I'm not really sure how to check that. We could replace them by their initial values maybe? Not sure that's a good idea @MichaelClerx ?

Reply:
I've added some extra explanations explaining the limitation, my tests show it only happens in the synthetic tests, not in any real models.